### PR TITLE
Add collections to image

### PR DESCRIPTION
**Pull request by jamesgorrie**
Looking to add slightly more complex data to Dynamo.

Unfortunately adding JSON isn't possible at the attribute level, but only at the `Item` level. You also loose a bit of the programatic side of things, and it turns out it's only an abstraction over the `ValueMap` system.

Potentially we could add another table that deals with collections alone - but it feels like it belongs here.

__To think about__
- [x] What do we do with duplicates, at the moment we just accumulatively add.
- [ ] Should this be a separate system? If metadata editor goes down - should this go down with it?

### CODE REVIEW

**Comment by jamesgorrie** [comment]
Waiting till Monday to merge #1457 that'll make this clearer

**Comment by jamesgorrie** [comment]
A full `/metadata/:id` response would look like:

```JSON
{
    "data": {
        "archived": {
            "uri": "https://media-metadata.local.dev-gutools.co.uk/metadata/ce83af9251f41587297d6790edf9bdd2514804da/archived",
            "data": false
        },
        "labels": {
            "uri": "https://media-metadata.local.dev-gutools.co.uk/metadata/ce83af9251f41587297d6790edf9bdd2514804da/labels",
            "data": []
        },
        "metadata": {
            "uri": "https://media-metadata.local.dev-gutools.co.uk/metadata/ce83af9251f41587297d6790edf9bdd2514804da/metadata",
            "data": {
                "keywords": []
            },
            "actions": [
                {
                    "name": "set-from-usage-rights",
                    "href": "https://media-metadata.local.dev-gutools.co.uk/metadata/ce83af9251f41587297d6790edf9bdd2514804da/metadata/set-from-usage-rights",
                    "method": "POST"
                }
            ]
        },
        "usageRights": {
            "uri": "https://media-metadata.local.dev-gutools.co.uk/metadata/ce83af9251f41587297d6790edf9bdd2514804da/usage-rights"
        },
        "collections": {
            "uri": "https://media-metadata.local.dev-gutools.co.uk/metadata/ce83af9251f41587297d6790edf9bdd2514804da/collections",
            "data": [
                {
                    "uri": "https://media-metadata.local.dev-gutools.co.uk/metadata/ce83af9251f41587297d6790edf9bdd2514804da/collections/g2/art/24%2F7",
                    "data": {
                        "pathId": "g2/art/24%2F7",
                        "path": [
                            "g2",
                            "art",
                            "24/7"
                        ],
                        "actionData": {
                            "author": "james.gorrie@guardian.co.uk",
                            "date": "2015-11-19T00:00:00.000+00:00"
                        }
                    }
                }
            ]
        }
    }
}
```

File: common-lib/src/main/scala/com/gu/mediaservice/lib/aws/DynamoDB.scala [hunk 1]
```diff
@@ -129,14 +137,17 @@ class DynamoDB(credentials: AWSCredentials, region: Region, tableName: String) {
         new ValueMap().withMap(":value", valueMapWithNullForEmptyString(value))
     )
 
-  def setDelete(id: String, key: String, value: String)
-               (implicit ex: ExecutionContext): Future[JsObject] =
+  def setObjAdd[T](id: String, listName: String, value: T)
+                  (implicit ex: ExecutionContext, tjs: Writes[T]): Future[JsObject] = {
+    val json = Json.toJson(value).as[JsObject]
+    val valueMap = DynamoDB.jsonToValueMap(json)
+    val e = s"SET $listName = list_append($listName, :value)"
```
Comment (jamesgorrie) [inline common-lib/src/main/scala/com/gu/mediaservice/lib/aws/DynamoDB.scala]: `ADD` has no affect on Lists.

File: common-lib/src/main/scala/com/gu/mediaservice/lib/collections/CollectionsManager.scala [hunk 2]
```diff
@@ -0,0 +1,14 @@
+package com.gu.mediaservice.lib.collections
+
+import com.gu.mediaservice.model.Collection
+
+object CollectionsManager {
+  def onlyLatest(collections: List[Collection]): List[Collection] = {
```
Comment (jamesgorrie) [inline common-lib/src/main/scala/com/gu/mediaservice/lib/collections/CollectionsManager.scala]: Been talking to @akash1810 about this - so thought it's worth elaborating:

When storing collections in DynamoDB - we do it as simply as append the collection to the list. This can land up with two collections, with different `actionsData` attached. i.e.
```JSON
{
    "collections": [
        {
            "pathId": "g2",
            "path": [
                "g2"
            ],
            "actionData": {
                "author": "test",
                "date": "2015-11-23T10: 48: 59.475+00: 00"
            }
        },
        {
            "pathId": "g2/24%2F7",
            "path": [
                "g2",
                "24/7"
            ],
            "actionData": {
                "author": "test",
                "date": "2015-11-23T10: 57: 48.505+00: 00"
            }
        },
        {
            "pathId": "g2",
            "path": [
                "g2"
            ],
            "actionData": {
                "author": "test",
                "date": "2015-11-23T11: 28: 24.743+00: 00"
            }
        }
    ]
}
```

When something is rendered, and what we store in the ES record would be the deduped to latest version of the list, only Dynamo contains the whole trail (which could be useful).

i.e.
```JSON
{
    "collections": [
        {
            "pathId": "g2/24%2F7",
            "path": [
                "g2",
                "24/7"
            ],
            "actionData": {
                "author": "test",
                "date": "2015-11-23T10: 57: 48.505+00: 00"
            }
        },
        {
            "pathId": "g2",
            "path": [
                "g2"
            ],
            "actionData": {
                "author": "test",
                "date": "2015-11-23T11: 28: 24.743+00: 00"
            }
        }
    ]
}
```

TL;DR
The response from the metadata editor and ES record contain the deduped version, but dynamo contains the full record.

Comment (akash1810) [inline common-lib/src/main/scala/com/gu/mediaservice/lib/collections/CollectionsManager.scala]: Any reason to not make `actionData` a list too? So that we store:

```json
{
    "collections": [
        {
            "pathId": "g2",
            "path": [
                "g2"
            ],
            "actionData": [{
                "author": "test",
                "date": "2015-11-23T10: 48: 59.475+00: 00"
              }, {
                "author": "test",
                "date": "2015-11-23T11: 28: 24.743+00: 00"
            }]
        },
        {
            "pathId": "g2/24%2F7",
            "path": [
                "g2",
                "24/7"
            ],
            "actionData": [{
                "author": "test",
                "date": "2015-11-23T10: 57: 48.505+00: 00"
            }]
        }
    ]
}
```

The logic to create/update would become more complex, but might be worth considering...

Comment (jamesgorrie) [inline common-lib/src/main/scala/com/gu/mediaservice/lib/collections/CollectionsManager.scala]: I was thinking this - but you do land up wondering if that should then exist on `collections`. e.g. 
```JSON
{
  "collections": {
    "data": [],
    "actionData": [
      { "action": "edit", "date": "...", "author": "me@you.com" },
      ...
    ]
  }
}
```

Comment (jamesgorrie) [inline common-lib/src/main/scala/com/gu/mediaservice/lib/collections/CollectionsManager.scala]: So basically feeling if it was an audit trail, it would be more useful on the collection (of collections) rather than the singular item.

**Comment by theefer** [comment]
> Should this be a separate system? If metadata editor goes down - should this go down with it?

I'd argue it should be. We could move later, possibly, but worth keeping in mind how it is exposed and stored in ES and the media-api.

**Comment by theefer** [comment]
Actually raises the question whether it should be stored in the same dynamo table as metadata overrides/edits...?

**Comment by jamesgorrie** [comment]
> Actually raises the question whether it should be stored in the same dynamo table as metadata overrides/edits

This should definitely be a new service then - which means this PR is wrong (although the dynamo work which was the hard bit).
